bridgectrl/merkletree.go

```diff
@@ -142,28 +136,47 @@ func (mt *MerkleTree) addLeaf(ctx context.Context, leaf [KeyLen]byte) error {
 		return err
 	}
 
+	dbTx, err := mt.store.BeginDBTransaction(ctx)
+	if err != nil {
+		return err
+	}
+
+	var leaves [][][]byte
 	for h := uint8(0); h < mt.height; h++ {
 		if index&(1<<h) > 0 {
-			parent = Hash(siblings[h], cur)
-			err := mt.store.Set(ctx, parent[:], [][]byte{siblings[h][:], cur[:]}, nil)
-			if err != nil {
-				return err
-			}
+			var child [KeyLen]byte
+			copy(child[:], cur[:])
+			parent := Hash(siblings[h], child)
+			cur = parent
+			leaves = append(leaves, [][]byte{parent[:], siblings[h][:], child[:]})
 		} else {
-			parent = Hash(cur, siblings[h])
-			err := mt.store.Set(ctx, parent[:], [][]byte{cur[:], siblings[h][:]}, nil)
-			if err != nil {
-				return err
-			}
+			var child [KeyLen]byte
+			copy(child[:], cur[:])
+			parent := Hash(child, siblings[h])
+			cur = parent
+			leaves = append(leaves, [][]byte{parent[:], child[:], siblings[h][:]})
 		}
-		cur = parent
 	}
-
-	// Set the root value
+	// Set the root value and leaves
 	mt.root = cur
 	mt.count++
-	err = mt.store.SetRoot(ctx, cur[:], mt.count, mt.network, nil)
-	return err
+	rootID, err := mt.store.SetRoot(ctx, cur[:], mt.count, mt.network, dbTx)
+	if err != nil {
+		if rollbackErr := mt.store.Rollback(ctx, dbTx); rollbackErr != nil {
+			return rollbackErr
+		}
+		return err
+	}
+	for _, leaf := range leaves {
+		err := mt.store.Set(ctx, leaf[0], [][]byte{leaf[1], leaf[2]}, rootID, dbTx)
+		if err != nil {
+			if rollbackErr := mt.store.Rollback(ctx, dbTx); rollbackErr != nil {
+				return rollbackErr
+			}
+			return err
+		}
+	}
```

Comment:
Instead of inserting the leafs one by one, we could try to insert all the leaves at ones.
https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-insert-multiple-rows/

Reply:
we will do, you want it in this PR?

Reply:
You can create an issue for ir, this pr is big enough

Reply:
we already have https://github.com/0xPolygonHermez/zkevm-bridge-service/issues/284